File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length, maxi = 0;
        Stack<Integer> st = new Stack<>();
        for (int i = 0; i <= n; i++) {
            while (!st.isEmpty() && (i == n || heights[st.peek()] >= heights[i])) {
                int height = heights[st.pop()];
                int width;
                if (st.isEmpty()) width = i;
                else width = i - st.peek() - 1;
                maxi = Math.max(maxi, height * width);
            }
            st.push(i);
        }
        return maxi;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        stack<int> st;
        
        for (int i = 0, n = heights.size(); i <= n; i++) {
            while (!st.empty() && (i == n || heights[st.top()] >= heights[i])) {
                int l = heights[st.top()]; st.pop();
                int b = st.empty() ? i : i - st.top() - 1;
                res = max(res, l * b);
            }
            st.push(i);
        }
        
        return res;
    }
};